src/fcntl/statx.rs

```diff
@@ -0,0 +1,371 @@
+    dirfd: Option<RawFd>,
+    pathname: &P,
+    flags: Flags,
+    mask: Mask,
+) -> Result<Statx> {
+    let dirfd = dirfd.unwrap_or(libc::AT_FDCWD);
+
+    let mut dst = mem::MaybeUninit::uninit();
+    let res = pathname.with_nix_path(|cstr| unsafe {
+        libc::statx(
+            dirfd,
+            cstr.as_ptr(),
+            flags.bits() as libc::c_int,
+            mask.bits() as libc::c_uint,
+            dst.as_mut_ptr(),
+        )
+    })?;
+
+    Errno::result(res)?;
+
+    Ok(Statx {
+        inner: unsafe { dst.assume_init() },
+    })
+}
+
+/// Get stats of specified file descriptor (not nesessarily directory) without resolving the path.
+/// Automaitcally sets the `AT_EMPTY_PATH` flag.
+pub fn statx_without_path(dirfd: RawFd, mut flags: Flags, mask: Mask) -> Result<Statx> {
+    flags.set(Flags::AT_EMPTY_PATH, true);
+    let mut dst = mem::MaybeUninit::uninit();
+    let res = unsafe {
+        libc::statx(
+            dirfd,
+            b"\0" as *const u8 as *const libc::c_char,
+            flags.bits() as libc::c_int,
+            mask.bits() as libc::c_uint,
+            dst.as_mut_ptr(),
+        )
+    };
+
+    Errno::result(res)?;
+
+    Ok(Statx {
+        inner: unsafe { dst.assume_init() },
+    })
+}
+
+#[derive(Debug,Copy,Clone)]
+pub struct Statx {
+    pub inner: libc::statx,
+}
+
+impl Statx {
+    /// Retrieve file type, if it has been returned by kernel
+    pub fn r#type(&self) -> Option<SFlag> {
+        if Mask::STATX_TYPE.bits() & self.inner.stx_mask > 0 {
+            Some(SFlag::from_bits_truncate(self.inner.stx_mode as mode_t & SFlag::S_IFMT.bits()))
+        } else {
+            None
+        }
+    }
+    
+    /// Retrieve file mode, if it has been returned by kernel.
+    pub fn mode(&self) -> Option<Mode> {
+        if Mask::STATX_MODE.bits() & self.inner.stx_mask > 0 {
+            Some(Mode::from_bits_truncate(self.inner.stx_mode as mode_t))
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve number of hard links, if it has been returned by kernel.
+    pub fn nlinks(&self) -> Option<u32> {
+        if Mask::STATX_NLINK.bits() & self.inner.stx_mask > 0 {
+            Some(self.inner.stx_nlink)
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve uid (owner, user ID), if it has been returned by kernel.
+    pub fn uid(&self) -> Option<Uid> {
+        if Mask::STATX_NLINK.bits() & self.inner.stx_mask > 0 {
+            Some(Uid::from_raw(self.inner.stx_uid))
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve gid (group ID), if it has been returned by kernel.
+    pub fn gid(&self) -> Option<Gid> {
+        if Mask::STATX_NLINK.bits() & self.inner.stx_mask > 0 {
+            Some(Gid::from_raw(self.inner.stx_uid))
+        } else {
+            None
+        }
+    }
+
+    /// Retrieve file access time, if it has been returned by kernel
+    pub fn atime(&self) -> Option<libc::statx_timestamp> {
```

Comment:
Returning the raw libc type isn't very useful.  Is it identical to a `struct timespec`?  If so, I suggest you convert it into a `nix::sys::time::TimeSpec`.

Reply:
Done, also added the appropriate `From` implementation within the `statx` module. Or should it be moved to the place where other `TimeSpec` `From` implementations live with a duplicated `#[cfg]` line?